### DISCUSSION TITLE
Smart Sources

### DISCUSSION THREAD

**Discussion by brandonkelly** [craftcms/cms#8423] [Ideas] [answered]
Element indexes should have the ability to define Smart Sources, with much more control over criteria params than the main index view offers. Similar to things like this:

<img width="583" alt="screen shot 2017-04-13 at 2 21 44 pm" src="https://cloud.githubusercontent.com/assets/47792/25025021/6cbe4b9e-2055-11e7-9488-9ee87ab1fbd0.png">

They should be defined within the CustomizeSourcesModal that opens when you click the gear icon at the bottom of element source lists.

**Comment by sjelfull** [comment]
Very nice. Just to clarify, would a Smart Source be in addition to the main index?

Ideally, I think this filtering should be available everywhere, but it would make sense to have Smart Sources, which is basically saved filtering?

**Comment by brandonkelly** [comment]
Yes, they would show up wherever the element type’s sources are listed:

- main element indexes
- relational fields’ Sources settings
- element selection modals

**Comment by brandonkelly** [comment]
This would also resolve #838 and #1024.

**Comment by sjelfull** [comment]
But can you do on the fly filtering in addition to the Smart Source? Ie. you don't have create a Smart Source just to filter, right?

If you create a first-party UI for setting up rules, it would be awesome if we can use that in plugins to create our own rule UIs.

**Comment by brandonkelly** [comment]
We could make the UI reusable for on-the-fly filtering as well, yeah. And yes, all element types will be able to define their own filter options, etc., and plugins would be able to tweak them with an event, like how they can tweak most other element type things.

**Comment by AugustMiller** [comment]
Is it possible to conceptualize _all_ element indexes (i.e. Section tables) as _Smart Sources_?

For example, when creating a section, Craft might auto-generate a new _Smart Source_ for it (with the current behavior), but it could be honed further, deleted, etc.

**Comment by brandonkelly** [comment]
@AugustMiller That's an interesting idea, for sure! The only semi-issue I can think of is, it would mean that element types _must_ support Smart Sources if they want to show any sources at all. But that’s not necessarily a deal-breaker.

**Comment by elivz** [comment]
This would be amazing. One use case to add to the list is what I [describe here](https://craftcms.stackexchange.com/questions/22807/how-to-differentiate-between-positive-and-negative-numbers-in-search/22812#22812): currently there is no way to differentiate between "10" and "-10" in the entries list search, even with a number-type field. While we're at it, it would be nice to have the ability to sort by number-type custom fields (or text fields for that matter) in those lists.

**Comment by gojira32** [comment]
@brandonkelly Is there any movement on this one?

**Comment by brandonkelly** [comment]
Yep. On the list for Craft 4 (in dev).

**Comment by brandonkelly** [comment, marked as answer]
This has been implemented for Craft 4 🎉  (#10075)

**Comment by arentsen** [comment]
I love the new Element index filtering option in Craft 4! It would be great if I could make a filter like "Show me all articles created by [me]", [me] being the current user. Now it seems I can only select a specific user. I tried to enter some global variables in the fields, but that did not seem to work. Thanks.

**Reply by brandonkelly** [reply]
That’s going to be possible in Craft 4.4 thanks to #12679!

<img width="784" alt="A custom entry source’s “Entry Criteria” setting, set to pull in entries in a “News” section, with an “Article” entry type, and with an Author set to “{{currentUser.id}}”, a dynamic Twig template." src="https://user-images.githubusercontent.com/47792/220410171-c4a62333-0d47-42ef-b6bd-2ed9401a50df.png">
